### DISCUSSION TITLE
MVP Features

### DISCUSSION THREAD

**Discussion by gaqzi** [gaqzi/incident-reviewer#18] [Design]
The basic idea for the review is [in the mockup](https://github.com/gaqzi/incident-reviewer/blob/main/docs/images/incident-reviewer.excalidraw.svg) but I want to get some feedback on how I'm thinking about getting those various sections to work and what to call them.

I have been talking about "root causes" but also know there has been a move towards not using that word because it implies there is only one cause. Looking at some research done by others I've found "contributing and proximal causes" where proximal is "the closest thing" that was the reason (or what I would've previously called _the_ root cause), and contributing are other things that either helped make the incident possible or made it worse (we didn't know how to access a system, didn't know how to parse, the firewall wasn't configured correctly, etc.).

So the MVP features as I'm thinking of implementing them are:

## Basic information

Link to an existing report and provide the basic information we need to understand it. Assume it's keyed in/copied from the original report

Pieces of data
- Title
- Description
- Impact (orders lost, marketing budget wasted, etc.)
- Where (countries, regions, etc.)
- The root cause as described in the report (text, not classified)
- The trigger, as described in the report

## Known contributing and proximal causes

Provide classified/normalized/known "Contributing and proximal" causes with descriptions

- Select causes from list (hardcoded and with explanations)
- Why this cause applies
- Have a checkbox for proximal; there can be only one for each report. The idea is that while we can have many contributing causes (that made it worse or opened the gate for it to fail) this one alone is most directly the cause.
- Context around this cause (tags)
    - Where will these tags come from? Will we have a strict list from the start or allow adding any tag and then rely on admins to merge/alias as necessary? Let's start with free-form and then we'll decide later.
- Allow for reactions with why you react that way
- How to identify (command to run, symptoms you're seeing, etc. free text with votes)
- Allow for write-in causes so we can add new

## Known triggers
Provide classified/normalized/known  triggers with descriptions

- Select triggers from list
- Why this trigger applies
- Context around this trigger (tags)
- Allow for reactions with why you react that way
- How to identify this trigger (free text, intended to be specific to the incident, but can be part of finding the triggers in a list or find other matching later)
- Allow for write-in triggers so we can add new

## Learnings per incident

Lessons we should take from the full incident

- What (The feedback we should take from the system) [required]
- The thing we should look out for, the action to take, or what to avoid.
- I expect that calling out the action is different from identifying whether it applies in some cases, and it might need more context or "it depends" factors to make it truly useful.
- "What it depends on"/How to identify [optional]
    - In these circumstances you should apply this lesson
- Context (tags)
- Allow for reactions with why you react that way
## Action items

The actions we should take to not repeat this incident or to help detect or mitigate if we get it or similar in the future. Not all incidents are possible to avoid and mitigating them faster is a good outcome if that's what's available.

-  What (The action we need to take)
- Why this action
- Specify trigger + root cause where this action applies
- When (short/mid/long -term)
- Context (tags)
- Allow for voting up/down and with a reason for why you agree/disagree

## Known learnings

A repeat learning that maybe is discovered differently so elaborate on why it applies in this circumstance.

Known learnings (repeated) come with a writeup of "how to address it" with options and descriptions. Maybe linked to Confluence/engineering manifesto/etc. The point is to mark it and then help people find what to do about it. This way the team get suggestions for how to address it and doesn't have to do it themselves. 

- Select known learning from a list
- Why this learning applies in this case
- Reactions with reasons
- Allow for write-in learnings so we can add new

## Questions
As the tool is assumed to be used both solo and in meetings with 20+ attendees of various levels of experience.

Questions / Q&A that's attached to the incident report. As a way to allow people in the meeting to get clarifications on things they don't understand without having to unmute and ask. To help us drive towards more learnings and clarifications while retaining it between meetings.

Allow for answers to be marked as "correct" like Stack Overflow so they're bubbled to the top.

I think it's important that questions can be anonymous but we also need to avoid people being absolute dicks when they don't have their name attached. As a way to work around it I think each anonymous question should have a consistent ID attached that's visible to the admins and they can block frequent offenders from posting anonymously.

Admins can:

- Delete questions and mark why (duplicate, inflammatory, rude, etc.)

## Cross functional requirements

- Can use the tool with ~500pax at the same time and they get a near real-time experience of changes
    - Be able to stream small changes out to connected clients, but also to force full refetches as necessary
    - The plan is to build it with HTMX and use SSE for pushing changes out to connected clients. And then heavily rely on HTTP caching to make it performant.
- Simple to maintain by a backend dev who works on it sporadically, maintaining dependencies needs to be dead-simple and not break stuff __all the time__ 
- Fully tested with both browser, integration, and unit tests (d'uh)

**Comment by einhverfr** [comment]
A couple thoughts here.  Not sure how much of this needs to be in an MVP.

First, I think proximal cause  is similar to your trigger.  I.e. it is what was the last link in the chain (temporally) to break.

One key aspect of analysis here needs to be a division of threats and errors, and further looking at these as active or latent.

An error is a human operational error.  A threat is any systemic weakness or problem that exists.  Active means having an immediate, observable effect, and latent means having no noticeable effect yet.  The worst pairing is a mixture of latent and active problems (usually a latent threat and an active error).

Secondly if one wants to go with a swiss cheese approach, being able to add security barriers and why they failed (i.e. monitoring dashboards, human oversight, peer review of PRs, etc.  A drop down list of human error classifications would be a very good idea along with "technical fault" (bug, unfrastructure issue) and "technical threat" (design decision), and "organizational threat" (i.e. decision made as to what to tolerate, examples being "no on call for tier three service" etc).

I think I sent you a taxonomy of human error spreadsheet at some pojnt.  But if you don't have it, here would be my suggested dropdown values for human error classification with examples in parentheses):  Some of this may be outside what I sent though.

1.  Fatigue-related judgment failure (Fatigued engineer fails to see problem and escallate)
2.  Fatigue-related Error (typo in operation by fatigued engineer or missing step in repetitive task)
3.  Reversion under Stress (typo under stress, usually of a form that would more often be typed)
4.  Stress-related Fixation (overly focused on the wrong things under stress)
5.  Loss of situational awareness (Unable to properly understand the state of the system)
6.  Missed Steps under High Workload (too much going on and so some things missed)
7.  Other Memory Error. (missing step in a manual process, repeating a step by accident in a manual process)
8.  Other Execution Error (other typo, usually, in a command)
9.  Complacency (No longer following the rules or policies -- either the policies need to be changed or they need to be returned to)
10.  Technical Fault (Bug)
11.  Technical Threat (Design decision that didn't hold up)
12.  Organizational Threat (for example, not having on call for a tier three service and having a tier one service accidentally depend on it)
13. Failure to Escallate
14.  Other Coordination Breakdown

Longer term the Hazardous Attitudes could be added.

**Comment by einhverfr** [comment]
Just as a note for the Hazardous Attitudes -- these are usually in our industry organizational patters that develop.  The FAA lists the following but this is definitely not MVP material:

1. Anti-authority (breakdown in respect for the rules)
2. Impulsivity (the desire to do something now)
3. Invulnerability (belief that bad things cannot happen here)
4.  Macho (the desire to prove to others that you can do it)
5.  Resignation (everything has gone bad, so what's the use?)

My sense is that our industry tends to do reasonably well on 3, 4, and 5 but we have problems of trust in process breaking down because we are bad at building working processes, and there can be a lot of pressure to "fix it now" which can lead to impulsivity.
